File: Week 5/3.RecursionBasics.java
/*
 * WEEK 5 - FUNCTIONS & RECURSION
 * Topic: Recursion Basics
 *
 * WHAT IS RECURSION?
 * A function that calls itself to solve a smaller version of the same problem.
 *
 * EVERY RECURSIVE FUNCTION NEEDS:
 * 1. BASE CASE  — the condition that stops recursion (prevents infinite loop)
 * 2. RECURSIVE CASE — the function calling itself with a smaller/simpler input
 *
 * THE CALL STACK:
 * Each recursive call adds a new frame to the call stack.
 * When the base case is reached, frames are popped in reverse order.
 *
 * COMMON MISTAKE: Forgetting the base case → StackOverflowError
 *
 * PROBLEMS COVERED:
 * 1. Print numbers from N down to 1
 * 2. Factorial (N!)
 * 3. Sum of first N natural numbers
 * 4. Power (base^exp)
 *
 * Time Complexity:
 * - Factorial / Sum: O(n) — n recursive calls
 * - Power (naive): O(exp)
 * - Power (fast exponentiation): O(log exp)
 *
 * Space Complexity: O(n) for the call stack (n frames)
 */

public class RecursionBasics {

    // 1. Print N down to 1
    // Base case: n == 0 → stop
    static void printDesc(int n) {
        if (n == 0) return;       // base case
        System.out.print(n + " ");
        printDesc(n - 1);         // recursive call with smaller n
    }

    // 2. Print 1 up to N (print AFTER the recursive call)
    static void printAsc(int n) {
        if (n == 0) return;
        printAsc(n - 1);
        System.out.print(n + " ");
    }

    // 3. Factorial: N! = N * (N-1)!
    // Base case: 0! = 1, 1! = 1
    // Recurrence: fact(n) = n * fact(n-1)
    static long factorial(int n) {
        if (n <= 1) return 1;     // base case
        return n * factorial(n - 1);
    }

    // 4. Sum of first N natural numbers
    // Recurrence: sum(n) = n + sum(n-1)
    static int sum(int n) {
        if (n == 0) return 0;
        return n + sum(n - 1);
    }

    // 5. Power (base^exp) — naive O(exp)
    static long power(long base, int exp) {
        if (exp == 0) return 1;
        return base * power(base, exp - 1);
    }

    // 6. Fast Power (base^exp) — O(log exp)
    // If exp is even: base^exp = (base^(exp/2))^2
    // If exp is odd:  base^exp = base * base^(exp-1)
    static long fastPower(long base, int exp) {
        if (exp == 0) return 1;
        if (exp % 2 == 0) {
            long half = fastPower(base, exp / 2);
            return half * half;
        } else {
            return base * fastPower(base, exp - 1);
        }
    }

    public static void main(String[] args) {
        System.out.print("Descending 5 to 1: ");
        printDesc(5);
        System.out.println();

        System.out.print("Ascending 1 to 5: ");
        printAsc(5);
        System.out.println();

        System.out.println("5! = " + factorial(5));   // 120
        System.out.println("10! = " + factorial(10)); // 3628800

        System.out.println("sum(10) = " + sum(10));   // 55

        System.out.println("2^10 = " + power(2, 10));          // 1024
        System.out.println("2^10 (fast) = " + fastPower(2, 10)); // 1024
        System.out.println("3^20 (fast) = " + fastPower(3, 20)); // 3486784401
    }
}

File: Week 5/rust/3.recursion_basics.rs
/*
 * WEEK 5 - RUST FUNCTIONS & RECURSION
 * Topic: Recursion Basics
 * File: 3.recursion_basics.rs
 *
 * CONCEPT:
 *  Rust supports recursion but does NOT guarantee tail-call optimisation
 *  (although LLVM may do it in release mode for tail-recursive forms).
 *  Stack size is OS-default (~8 MB on Linux); deep recursion may overflow.
 *
 * KEY POINTS:
 *  - Mutability for accumulators -- pass `&mut Vec` etc.
 *  - For very deep recursion, prefer loops or explicit stacks.
 */

fn print_desc(n: i32) {
    if n == 0 { return; }
    print!("{n} ");
    print_desc(n - 1);
}

fn print_asc(n: i32) {
    if n == 0 { return; }
    print_asc(n - 1);
    print!("{n} ");
}

fn factorial(n: u64) -> u64 {
    if n <= 1 { 1 } else { n * factorial(n - 1) }
}

fn sum_n(n: i64) -> i64 {
    if n == 0 { 0 } else { n + sum_n(n - 1) }
}

fn power(base: i64, exp: u32) -> i64 {
    if exp == 0 { 1 } else { base * power(base, exp - 1) }
}

fn fast_power(base: i64, exp: u32) -> i64 {
    if exp == 0 { return 1; }
    if exp % 2 == 0 {
        let h = fast_power(base, exp / 2);
        h * h
    } else {
        base * fast_power(base, exp - 1)
    }
}

fn main() {
    print!("Descending 5..1: "); print_desc(5); println!();
    print!("Ascending  1..5: "); print_asc(5);  println!();

    println!("5!  = {}", factorial(5));
    println!("10! = {}", factorial(10));

    println!("sum(10) = {}", sum_n(10));
    println!("2^10 = {}", power(2, 10));
    println!("2^10 (fast) = {}", fast_power(2, 10));
    println!("3^20 (fast) = {}", fast_power(3, 20));
}

/*
 * NOTES:
 *  - u64 holds factorial up to 20!; beyond that switch to u128 or num-bigint.
 *  - Default stack is ample for ~10000-frame depths; for more, use std::thread::Builder
 *    to set a custom stack size.
 */
